partner_identification_stock/models/stock_picking.py

```diff
@@ -0,0 +1,43 @@
+from odoo import _, fields, models
+from odoo.exceptions import UserError
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    def _compute_id_requirement(self):
+        for rec in self:
+            rec.require_id = rec.valid_id = False
+            for line in rec.move_lines:
+                if line.product_id.require_id_stock_ids:
+                    rec.require_id = True
+                ffl_lines = rec.partner_id.id_numbers.filtered(
+                    lambda i: i.category_id.id
+                    in line.product_id.require_id_stock_ids.ids
+                    and i.valid_until
+                )
+                if any(
+                    line.valid_until >= fields.Date.today()
+                    and line.status in ["open", "pending"]
+                    for line in ffl_lines
+                ):
+                    rec.valid_id = True
+
+    require_id = fields.Boolean(string="Require ID?", compute="_compute_id_requirement")
```

Comment:
Fields names should be more specific in order to avoid collision.

require_partner_identification and valid_partner_identification should be more convenient.